sql/core/src/main/scala/org/apache/spark/sql/execution/ExplainUtils.scala

```diff
@@ -0,0 +1,237 @@
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.catalyst.trees.TreeNodeTag
+
+object ExplainUtils {
+  /**
+   * Given a input physical plan, performs the following tasks.
+   *   1. Computes the operator id for current operator and records it in the operaror
+   *      by setting a tag.
+   *   2. Computes the whole stage codegen id for current operator and records it in the
+   *      operator by setting a tag.
+   *   3. Generate the two part explain output for this plan.
+   *      1. First part explains the operator tree with each operator tagged with an unique
+   *         identifier.
+   *      2. Second part explans each operator in a verbose manner.
+   *
+   * Note : This function skips over subqueries. They are handled by its caller.
+   *
+   * @param plan Input query plan to process
+   * @param append function used to append the explain output
+   * @startOperationID The start value of operation id. The subsequent operations will
+   *                  be assigned higher value.
+   *
+   * @return The last generated operation id for this input plan. This is to ensure we
+   *         always assign incrementing unique id to each operator.
+   *
+   */
+  private def processPlanSkippingSubqueries[T <: QueryPlan[T]](
+      plan: => QueryPlan[T],
+      append: String => Unit,
+      startOperatorID: Int): Int = {
+
+    // ReusedSubqueryExecs are skipped over
+    if (plan.isInstanceOf[BaseSubqueryExec]) {
+      return startOperatorID
+    }
+
+    val operationIDs = new mutable.ArrayBuffer[(Int, QueryPlan[_])]()
+    var currentOperatorID = startOperatorID
+    try {
+      currentOperatorID = generateOperatorIDs(plan, currentOperatorID, operationIDs)
+      generateWholeStageCodegenIds(plan)
+
+      QueryPlan.append(
+        plan,
+        append,
+        verbose = false,
+        addSuffix = false,
+        printOperatorId = true)
+
+      append("\n")
+      var i: Integer = 0
+      for ((opId, curPlan) <- operationIDs) {
+        append(curPlan.verboseStringWithOperatorId())
+      }
+    } catch {
+      case e: AnalysisException => append(e.toString)
+    }
+    currentOperatorID
+  }
+
+  /**
+   * Given a input physical plan, performs the following tasks.
+   *   1. Generates the explain output for the input plan excluding the subquery plans.
+   *   2. Generates the explain output for each subquery referenced in the plan.
+   */
+  def processPlan[T <: QueryPlan[T]](
+      plan: => QueryPlan[T],
+      append: String => Unit): Unit = {
+    try {
+      val subqueries = ArrayBuffer.empty[(SparkPlan, Expression, SparkPlan)]
+      var currentOperatorID = 0
+      currentOperatorID = processPlanSkippingSubqueries(plan, append, currentOperatorID)
+      getSubqueries(plan, subqueries)
+      var i = 0
+
+      // Process all the subqueries in the plan.
+      for (sub <- subqueries) {
+        if (i == 0) {
+          append("\n===== Subqueries =====\n\n")
+        }
+        i = i + 1
```

Comment:
nit:
```
if (subqueries.nonEmpty) append("\n===== Subqueries =====\n\n")
for (sub <- subqueries) ...
```
then we don't need the `var i = 0`

Reply:
@cloud-fan actually i need that variable to print the subquery number ? I print like :
```
subquery:1
....
subquery:2
....
```